### ISSUE TITLE
Evidence of bug in query and validate that warrants further investigation

### ISSUE THREAD

**Issue by bbakerman** [graphql-java/graphql-java#2740]
### Discussed in https://github.com/graphql-java/graphql-java/discussions/2739

<div type='discussions-op-text'>

<sup>Originally posted by **arjunballa** February 25, 2022</sup>
How to validate graphql query against the schema without execution?

I tried below methods and all failed.

@bbakerman  @andimarek  Any pointers?

**Schema:** No mutations defined
**Request:** mutation
**Method:** ParseAndValidate.validate
Returns it as `valid`. If I add mutation with dummy `FieldDefinition` it works. 

```
	@Test
	public void testValidateWithInvalidRequestUsingSchemaWithOnlyQueryAndInputWithMutation()
			throws SchemaProblem, FileNotFoundException {

		InputStream in = this.getClass().getResourceAsStream("/only-query.graphql");

		TypeDefinitionRegistry typeDefinitionRegistry = new SchemaParser().parse(in);

		GraphQLSchema schema = UnExecutableSchemaGenerator.makeUnExecutableSchema(typeDefinitionRegistry);

		String request = "mutation MyMutation { mymutation }";

		Document inputDocument = new Parser().parseDocument(request);

		List<ValidationError> errors = ParseAndValidate.validate(schema, inputDocument);

		assertEquals(1, errors.size());
	}

```
**Schema:** No mutations defined
**Request:** mutation
**Method:** ParseAndValidate.parseAndValidate
Returns it as `valid`. If I add mutation with dummy `FieldDefinition` it works. 
```
	@Test
	public void testParseAndValidateWithInvalidRequestUsingSchemaWithOnlyQueryAndInputWithMutation()
			throws SchemaProblem, FileNotFoundException {
		InputStream in = this.getClass().getResourceAsStream("/only-query.graphql");

		TypeDefinitionRegistry typeDefinitionRegistry = new SchemaParser().parse(in);

		GraphQLSchema schema = UnExecutableSchemaGenerator.makeUnExecutableSchema(typeDefinitionRegistry);

		String request = "mutation MyMutation { mymutation }";

		ExecutionInput input = ExecutionInput.newExecutionInput(request).build();

		ParseAndValidateResult result = ParseAndValidate.parseAndValidate(schema, input);

		assertEquals(1, result.getErrors().size());
	}
```
**Schema:** Has mutation with not null return type
**Request:** mutation
**Method:** ParseAndValidate.parseAndValidate
MockedWiringFactory is expecting the actual return.
**Error:** 
```
java.lang.AssertionError: expected same:<> was not:<[graphql.execution.MissingRootTypeException: 
Schema is not configured for mutations.]>
```

```
	@Test
	public void testMockedWiringFactoryWithValidRequest() throws SchemaProblem, FileNotFoundException {
		InputStream in = this.getClass().getResourceAsStream("/mutation-query.graphql");

		TypeDefinitionRegistry typeDefinitionRegistry = new SchemaParser().parse(in);

		RuntimeWiring runtimeWiring = RuntimeWiring.newRuntimeWiring().wiringFactory(new MockedWiringFactory()).build();

		SchemaGenerator schemaGenerator = new SchemaGenerator();

		GraphQLSchema graphQLSchema = schemaGenerator.makeExecutableSchema(typeDefinitionRegistry, runtimeWiring);

		GraphQL graphQL = GraphQL.newGraphQL(graphQLSchema).build();

		String request = "mutation MyMutation { mymutation }";

		ExecutionInput input = ExecutionInput.newExecutionInput(request).build();

		ExecutionResult executionResult = graphQL.execute(input);

		assertSame("", executionResult.getErrors().toString());

		assertEquals(0, executionResult.getErrors().size());
	}
```

</div>

**Comment by bbakerman** [comment]
only-query.graphqls

```
type Query {
        myquery : String!
}
```

mutation-query.graphqls

```
type Query {
        myquery : String!
}
type Mutation {
        mymutation : String!
}
```

**Comment by bbakerman** [comment]
Ahh I just replicated these in some local tests.  And yeah it because this check is not implemented as a ValidationRule (and hence  ParseAndValidate.validate would catch it) but rather is implemented later in the `graphql.execution.Execution#getOperationRootType` call that throws an exception if there is no root type for that operation type.

I can see that it ideally should be a validation rule. I will look into this more to see why it is like it is.  It might be just age (this is really really old code) and it was done there and forgotten about.

It wont validate on a full request but as you say it does not get caught by the ParseAndValidate and code

**Comment by bbakerman** [comment]
I made a failing branch showing this

https://github.com/graphql-java/graphql-java/pull/2741

**Comment by arjunballa** [comment]
@bbakerman Any chance it will be fixed soon? What release? Will it be merged back to 17.1?

**Comment by bbakerman** [comment]
I am not sure it will be fixed soon.  This is one validation case out a raft of validation cases and it is applied at the start of execution rather than ParseAndValidate stage - so its caught in effect.  I am asking others on why the code is like it is.  There might be a good reason on why this could not be a validation rule
